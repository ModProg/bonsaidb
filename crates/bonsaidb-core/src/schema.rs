mod collection;
mod names;
mod schematic;
mod summary;
/// Types for defining map/reduce-powered `View`s.
pub mod view;

pub use bonsaidb_macros::{Collection, Schema, View, ViewSchema};

pub use self::collection::{
    AsyncEntry, AsyncList, Collection, DefaultSerialization, Entry, EntryInsert, EntryUpdate,
    InsertError, List, Nameable, NamedCollection, NamedReference, SerializedCollection,
};
pub use self::names::{
    Authority, CollectionName, InvalidNameError, Name, Qualified, QualifiedName, SchemaName,
    ViewName,
};
pub use self::schematic::Schematic;
pub use self::summary::{CollectionSummary, SchemaSummary, ViewSummary};
pub use self::view::map::{Map, MappedValue, ViewMappedValue};
pub use self::view::{
    CollectionMapReduce, DefaultViewSerialization, MapReduce, ReduceResult, SerializedView, View,
    ViewMapResult, ViewSchema,
};
use crate::Error;

/// Defines a group of collections that are stored into a single database.
///
/// ## Deriving this trait
///
/// This trait can be derived rather than manually implemented:
///
/// ```rust
/// use bonsaidb_core::schema::{Collection, Schema};
/// use serde::{Deserialize, Serialize};
///
/// #[derive(Schema)]
/// #[schema(name = "MySchema", collections = [MyCollection])]
/// # #[schema(core = bonsaidb_core)]
/// pub struct MySchema;
///
/// #[derive(Serialize, Deserialize, Default, Collection)]
/// #[collection(name = "MyCollection")]
/// # #[collection(core = bonsaidb_core)]
/// pub struct MyCollection {
///     pub rank: u32,
///     pub score: f32,
/// }
/// ```
///
/// If you're publishing a schema for use in multiple projects, consider giving
/// the schema an `authority`, which gives your schema a namespace:
///
/// ```rust
/// use bonsaidb_core::schema::Schema;
///
/// #[derive(Schema)]
/// #[schema(name = "MySchema", authority = "khonsulabs", collections = [MyCollection])]
/// # #[schema(core = bonsaidb_core)]
/// pub struct MySchema;
///
/// # use serde::{Deserialize, Serialize};
/// # use bonsaidb_core::schema::Collection;
/// # #[derive(Serialize, Deserialize, Default, Collection)]
/// # #[collection(name = "MyCollection")]
/// # #[collection(core = bonsaidb_core)]
/// # pub struct MyCollection {
/// #    pub rank: u32,
/// #    pub score: f32,
/// # }
/// ```
pub trait Schema: Send + Sync + 'static {
    /// Returns the unique [`SchemaName`] for this schema.
    fn schema_name() -> SchemaName;

    /// Defines the `Collection`s into `schema`.
    fn define_collections(schema: &mut Schematic) -> Result<(), Error>;

    /// Retrieves the [`Schematic`] for this schema.
    fn schematic() -> Result<Schematic, Error> {
        Schematic::from_schema::<Self>()
    }
}

/// This implementation is for accessing databases when interacting with
/// collections isn't required. For example, accessing only the key-value store
/// or pubsub.
impl Schema for () {
    fn schema_name() -> SchemaName {
        SchemaName::new("", "")
    }

    fn define_collections(_schema: &mut Schematic) -> Result<(), Error> {
        Ok(())
    }
}

impl<T> Schema for T
where
    T: Collection + 'static,
{
    fn schema_name() -> SchemaName {
        SchemaName::from(Self::collection_name())
    }

    fn define_collections(schema: &mut Schematic) -> Result<(), Error> {
        schema.define_collection::<Self>()
    }
}
