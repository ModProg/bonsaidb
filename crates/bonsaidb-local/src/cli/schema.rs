use std::str::FromStr;

use bonsaidb_core::connection::{AsyncStorageConnection, StorageConnection};
use bonsaidb_core::schema::{
    CollectionName, InvalidNameError, SchemaName, SchemaSummary, ViewName,
};
use clap::Parser;

/// A schema query against a storage instance.
#[derive(Parser, Debug)]
pub struct Command {
    /// The name of the schema to query.
    pub name: Option<SchemaName>,

    /// The item in the schema to query.
    pub item: Option<CollectionOrView>,
}

impl Command {
    /// Executes the command on `storage`.
    pub fn execute<SC: StorageConnection>(self, storage: &SC) -> Result<(), crate::Error> {
        let schemas = storage.list_available_schemas()?;
        self.handle_schema_command(schemas)
    }

    /// Executes the command on `storage`.
    pub async fn execute_async<SC: AsyncStorageConnection>(
        self,
        storage: &SC,
    ) -> Result<(), crate::Error> {
        let schemas = storage.list_available_schemas().await?;
        self.handle_schema_command(schemas)
    }

    fn handle_schema_command(self, schemas: Vec<SchemaSummary>) -> Result<(), crate::Error> {
        if let Some(name) = self.name {
            let Some(schema) = schemas.into_iter().find(|s| s.name == name) else {
                return Err(crate::Error::Core(
                    bonsaidb_core::Error::SchemaNotRegistered(name),
                ));
            };

            if let Some(item) = self.item {
                match item {
                    CollectionOrView::View(view) => {
                        let Some(collection) = schema.collection(&view.collection) else {
                            return Err(crate::Error::Core(
                                bonsaidb_core::Error::CollectionNotFound,
                            ));
                        };
                        let Some(view) = collection.view(&view) else {
                            return Err(crate::Error::Core(bonsaidb_core::Error::ViewNotFound));
                        };
                        println!("Version: {}", view.version);
                        println!("Policy: {}", view.policy);
                    }
                    CollectionOrView::Collection(collection) => {
                        let Some(collection) = schema.collection(&collection) else {
                            return Err(crate::Error::Core(
                                bonsaidb_core::Error::CollectionNotFound,
                            ));
                        };
                        let mut views = collection.views().collect::<Vec<_>>();
                        views.sort_by(|v1, v2| v1.name.cmp(&v2.name));
                        for view in views {
                            println!("{}", view.name);
                        }
                    }
                }
            } else {
                print_collection_list(&schema);
            }
        } else if let Some(item) = self.item {
            eprintln!("missing `schema` for inspecting {item:?}");
            std::process::exit(-1);
        } else {
            print_schema_list(schemas);
        }
        Ok(())
    }
}

fn print_schema_list(mut schemas: Vec<SchemaSummary>) {
    schemas.sort_by(|s1, s2| s1.name.cmp(&s2.name));

    for schema in schemas {
        println!("{}", schema.name);
    }
}

fn print_collection_list(schema: &SchemaSummary) {
    let mut collections = schema.collections().collect::<Vec<_>>();
    collections.sort_by(|c1, c2| c1.name.cmp(&c2.name));

    for collection in collections {
        println!("{}", collection.name);
    }
}

/// A name that is either a [`CollectionName`] or [`ViewName`].
#[derive(Debug, Clone)]
pub enum CollectionOrView {
    /// A view name.
    View(ViewName),
    /// A collection name.
    Collection(CollectionName),
}

impl FromStr for CollectionOrView {
    type Err = InvalidNameError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Ok(view) = ViewName::from_str(s) {
            Ok(Self::View(view))
        } else {
            CollectionName::from_str(s).map(Self::Collection)
        }
    }
}
