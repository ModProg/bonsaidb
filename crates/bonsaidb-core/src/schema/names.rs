use std::borrow::Cow;
use std::fmt::{Debug, Display, Write};
use std::ops::Deref;
use std::str::FromStr;
use std::sync::Arc;

use serde::{Deserialize, Serialize};

/// A schema name. Cloning is inexpensive.
#[derive(Hash, PartialEq, Eq, Deserialize, Serialize, Debug, Clone, Ord, PartialOrd)]
#[serde(try_from = "String")]
#[serde(into = "String")]
pub struct Name {
    name: Arc<Cow<'static, str>>,
    needs_escaping: bool,
}

/// A name was unable to e parsed.
#[derive(thiserror::Error, Debug, Serialize, Deserialize, Clone)]
#[error("invalid name: {0}")]
pub struct InvalidNameError(pub String);

impl Name {
    /// Creates a new name.
    pub fn new<T: Into<Self>>(contents: T) -> Self {
        contents.into()
    }

    /// Parses a name that was previously encoded via [`Self::encoded()`].
    ///
    /// # Errors
    ///
    /// Returns [`InvalidNameError`] if the name contains invalid escape
    /// sequences.
    pub fn parse_encoded(encoded: &str) -> Result<Self, InvalidNameError> {
        let mut bytes = encoded.bytes();
        let mut decoded = Vec::with_capacity(encoded.len());
        while let Some(byte) = bytes.next() {
            if byte == b'_' {
                if let (Some(high), Some(low)) = (bytes.next(), bytes.next()) {
                    if let Some(byte) = hex_chars_to_byte(high, low) {
                        decoded.push(byte);
                        continue;
                    }
                }
                return Err(InvalidNameError(encoded.to_string()));
            }

            decoded.push(byte);
        }

        String::from_utf8(decoded)
            .map(Self::from)
            .map_err(|_| InvalidNameError(encoded.to_string()))
    }

    /// Returns an encoded version of this name that contains only alphanumeric
    /// ASCII, underscore, and hyphen.
    #[must_use]
    pub fn encoded(&self) -> String {
        format!("{self:#}")
    }
}

impl From<Cow<'static, str>> for Name {
    fn from(value: Cow<'static, str>) -> Self {
        let needs_escaping = !value
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'-');
        Self {
            name: Arc::new(value),
            needs_escaping,
        }
    }
}

impl From<&'static str> for Name {
    fn from(value: &'static str) -> Self {
        Self::from(Cow::Borrowed(value))
    }
}

impl From<String> for Name {
    fn from(value: String) -> Self {
        Self::from(Cow::Owned(value))
    }
}

#[allow(clippy::from_over_into)] // the auto into impl doesn't work with serde(into)
impl Into<String> for Name {
    fn into(self) -> String {
        self.name.to_string()
    }
}

impl Display for Name {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if f.alternate() && self.needs_escaping {
            for byte in self.name.bytes() {
                if byte.is_ascii_alphanumeric() || byte == b'-' {
                    f.write_char(byte as char)?;
                } else {
                    // Encode the byte as _FF
                    f.write_char('_')?;
                    f.write_char(nibble_to_hex_char(byte >> 4))?;
                    f.write_char(nibble_to_hex_char(byte & 0xF))?;
                }
            }
            Ok(())
        } else {
            Display::fmt(&self.name, f)
        }
    }
}

const fn nibble_to_hex_char(nibble: u8) -> char {
    let ch = match nibble {
        0..=9 => b'0' + nibble,
        _ => b'a' + nibble - 10,
    };
    ch as char
}

const fn hex_chars_to_byte(high_nibble: u8, low_nibble: u8) -> Option<u8> {
    match (
        hex_char_to_nibble(high_nibble),
        hex_char_to_nibble(low_nibble),
    ) {
        (Some(high_nibble), Some(low_nibble)) => Some(high_nibble << 4 | low_nibble),
        _ => None,
    }
}

const fn hex_char_to_nibble(nibble: u8) -> Option<u8> {
    let ch = match nibble {
        b'0'..=b'9' => nibble - b'0',
        b'a'..=b'f' => nibble - b'a' + 10,
        _ => return None,
    };
    Some(ch)
}

impl AsRef<str> for Name {
    fn as_ref(&self) -> &str {
        self.name.as_ref()
    }
}

/// The owner of a schema item. This should represent the company, group, or
/// individual that created the item in question. This value is used for
/// namespacing. Changing this after values are in use is not supported without
/// manual migrations at this time.
#[derive(Hash, PartialEq, Eq, Deserialize, Serialize, Debug, Clone, Ord, PartialOrd)]
#[serde(transparent)]
pub struct Authority(Name);

impl From<Cow<'static, str>> for Authority {
    fn from(value: Cow<'static, str>) -> Self {
        Self::from(Name::from(value))
    }
}

impl From<&'static str> for Authority {
    fn from(value: &'static str) -> Self {
        Self::from(Cow::Borrowed(value))
    }
}

impl From<String> for Authority {
    fn from(value: String) -> Self {
        Self::from(Cow::Owned(value))
    }
}

impl From<Name> for Authority {
    fn from(value: Name) -> Self {
        Self(value)
    }
}

impl Display for Authority {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl AsRef<str> for Authority {
    fn as_ref(&self) -> &str {
        self.0.as_ref()
    }
}

/// A namespaced name.
#[derive(Hash, PartialEq, Eq, Deserialize, Serialize, Debug, Clone, Ord, PartialOrd)]
pub struct QualifiedName {
    /// The authority that defines this name.
    pub authority: Authority,

    /// The name, unique within `authority`.
    pub name: Name,
}

impl Display for QualifiedName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.authority, f)?;
        f.write_char('.')?;
        Display::fmt(&self.name, f)
    }
}

impl FromStr for QualifiedName {
    type Err = InvalidNameError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse_encoded(s)
    }
}

/// Functions for creating qualified names
pub trait Qualified: Display + Sized {
    /// Creates a name that is not meant to be shared with other developers or
    /// projects.
    #[must_use]
    fn private<N: Into<Name>>(name: N) -> Self {
        Self::new(Authority::from("private"), name)
    }

    /// Creates a new qualified name.
    #[must_use]
    fn new<A: Into<Authority>, N: Into<Name>>(authority: A, name: N) -> Self;

    /// Parses a schema name that was previously encoded via
    /// [`Self::encoded()`].
    ///
    /// # Errors
    ///
    /// Returns [`InvalidNameError`] if the name contains invalid escape
    /// sequences or contains more than two periods.
    fn parse_encoded(schema_name: &str) -> Result<Self, InvalidNameError> {
        let mut parts = schema_name.split('.');
        if let (Some(authority), Some(name), None) = (parts.next(), parts.next(), parts.next()) {
            let authority = Name::parse_encoded(authority)?;
            let name = Name::parse_encoded(name)?;

            Ok(Self::new(authority, name))
        } else {
            Err(InvalidNameError(schema_name.to_string()))
        }
    }

    /// Encodes this schema name such that the authority and name can be
    /// safely parsed using [`Self::parse_encoded`].
    #[must_use]
    fn encoded(&self) -> String {
        format!("{self:#}")
    }
}

impl Qualified for QualifiedName {
    fn new<A: Into<Authority>, N: Into<Name>>(authority: A, name: N) -> Self {
        Self {
            authority: authority.into(),
            name: name.into(),
        }
    }
}

/// The name of a [`Schema`](super::Schema).
#[derive(Hash, PartialEq, Eq, Deserialize, Serialize, Debug, Clone, Ord, PartialOrd)]
#[serde(transparent)]
pub struct SchemaName(pub(crate) QualifiedName);

impl Qualified for SchemaName {
    fn new<A: Into<Authority>, N: Into<Name>>(authority: A, name: N) -> Self {
        Self(QualifiedName::new(authority, name))
    }
}

impl Display for SchemaName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl Deref for SchemaName {
    type Target = QualifiedName;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl From<CollectionName> for SchemaName {
    fn from(name: CollectionName) -> Self {
        Self(name.0)
    }
}

impl FromStr for SchemaName {
    type Err = InvalidNameError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse_encoded(s)
    }
}

/// The namespaced name of a [`Collection`](super::Collection).
#[derive(Hash, PartialEq, Eq, Deserialize, Serialize, Debug, Clone, Ord, PartialOrd)]
#[serde(transparent)]
pub struct CollectionName(pub(crate) QualifiedName);

impl Qualified for CollectionName {
    fn new<A: Into<Authority>, N: Into<Name>>(authority: A, name: N) -> Self {
        Self(QualifiedName::new(authority, name))
    }
}

impl Display for CollectionName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl Deref for CollectionName {
    type Target = QualifiedName;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl FromStr for CollectionName {
    type Err = InvalidNameError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse_encoded(s)
    }
}

/// The name of a [`View`](super::View).
#[derive(Hash, PartialEq, Eq, Deserialize, Serialize, Debug, Clone, PartialOrd, Ord)]
pub struct ViewName {
    /// The name of the collection that contains this view.
    pub collection: CollectionName,
    /// The name of this view.
    pub name: Name,
}

impl ViewName {
    /// Creates a new view name.
    pub fn new<
        C: TryInto<CollectionName, Error = InvalidNameError>,
        N: TryInto<Name, Error = InvalidNameError>,
    >(
        collection: C,
        name: N,
    ) -> Result<Self, InvalidNameError> {
        let collection = collection.try_into()?;
        let name = name.try_into()?;
        Ok(Self { collection, name })
    }
}

impl Display for ViewName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.collection, f)?;
        f.write_char('.')?;
        Display::fmt(&self.name, f)
    }
}

impl FromStr for ViewName {
    type Err = InvalidNameError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (first, view_name) = s
            .rsplit_once('.')
            .ok_or_else(|| InvalidNameError(s.to_string()))?;

        let collection = first.parse()?;
        let name = Name::parse_encoded(view_name)?;
        Ok(Self { collection, name })
    }
}

#[test]
fn name_escaping_tests() {
    const VALID_CHARS: &str = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-";
    const INVALID_CHARS: &str = "._hello\u{1F680}";
    const ESCAPED_INVALID: &str = "_2e_5fhello_f0_9f_9a_80";
    assert_eq!(Name::new(VALID_CHARS).to_string(), VALID_CHARS);
    assert_eq!(Name::new(INVALID_CHARS).to_string(), INVALID_CHARS);
    assert_eq!(Name::new(INVALID_CHARS).encoded(), ESCAPED_INVALID);
    assert_eq!(
        Name::parse_encoded(ESCAPED_INVALID).unwrap(),
        Name::new(INVALID_CHARS)
    );
    Name::parse_encoded("_").unwrap_err();
    Name::parse_encoded("_0").unwrap_err();
    Name::parse_encoded("_z").unwrap_err();
    Name::parse_encoded("_0z").unwrap_err();
}

#[test]
fn joined_names_tests() {
    const INVALID_CHARS: &str = "._hello\u{1F680}.._world\u{1F680}";
    const ESCAPED_INVALID: &str = "_2e_5fhello_f0_9f_9a_80._2e_5fworld_f0_9f_9a_80";
    let collection = CollectionName::parse_encoded(ESCAPED_INVALID).unwrap();
    assert_eq!(collection.to_string(), INVALID_CHARS);
    assert_eq!(collection.encoded(), ESCAPED_INVALID);

    let schema_name = SchemaName::parse_encoded(ESCAPED_INVALID).unwrap();
    assert_eq!(schema_name.to_string(), INVALID_CHARS);
    assert_eq!(schema_name.encoded(), ESCAPED_INVALID);
}
