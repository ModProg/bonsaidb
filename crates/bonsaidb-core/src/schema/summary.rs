use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::schema::{CollectionName, SchemaName, Schematic, ViewName};

/// A summary of a [`Schema`](crate::schema::Schema)/[`Schematic`].
///
/// This type is a serializable summary of a [`Schematic`] and is the result of
/// [`StorageConnection::list_available_schemas`](crate::connection::StorageConnection::list_available_schemas)/[`AsyncStorageConnection::list_available_schemas`](crate::connection::AsyncStorageConnection::list_available_schemas).
/// It can be used to query information stored in BonsaiDb without needing
/// access to the Rust types.
#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
pub struct SchemaSummary {
    /// The name of the [`Schema`](crate::schema::Schema) this summary is of.
    pub name: SchemaName,
    collections: HashMap<CollectionName, CollectionSummary>,
}

impl SchemaSummary {
    /// Returns the summary of named collection, if the schema contains it.
    #[must_use]
    pub fn collection(&self, name: &CollectionName) -> Option<&CollectionSummary> {
        self.collections.get(name)
    }

    /// Returns an iterator over all collections contained in this schema.
    pub fn collections(&self) -> impl Iterator<Item = &CollectionSummary> {
        self.collections.values()
    }
}

impl<'a> From<&'a Schematic> for SchemaSummary {
    fn from(schematic: &'a Schematic) -> Self {
        let mut summary = Self {
            name: schematic.name.clone(),
            collections: HashMap::new(),
        };

        for collection_name in schematic.collections() {
            let collection = summary
                .collections
                .entry(collection_name.clone())
                .or_insert_with(|| CollectionSummary {
                    name: collection_name.clone(),
                    views: HashMap::new(),
                });
            for view in schematic.views_in_collection(collection_name) {
                let name = view.view_name();
                collection.views.insert(
                    name.clone(),
                    ViewSummary {
                        name,
                        lazy: view.lazy(),
                        unique: view.unique(),
                        version: view.version(),
                    },
                );
            }
        }

        summary
    }
}

/// A summary of a [`Collection`](crate::schema::Collection).
#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
pub struct CollectionSummary {
    /// The name of the [`Collection`](crate::schema::Collection) this is a summary of.
    pub name: CollectionName,
    views: HashMap<ViewName, ViewSummary>,
}

impl CollectionSummary {
    /// Returns the summary of the named view, if it is contained in this collection.
    #[must_use]
    pub fn view(&self, name: &ViewName) -> Option<&ViewSummary> {
        self.views.get(name)
    }

    /// Returns an iterator over all summaries of views in this collection.
    pub fn views(&self) -> impl Iterator<Item = &ViewSummary> {
        self.views.values()
    }
}

/// A summary of a [`ViewSchema`](crate::schema::ViewSchema).
#[derive(Debug, Eq, PartialEq, Serialize, Deserialize, Clone)]
pub struct ViewSummary {
    /// The name of the [`ViewSchema`](crate::schema::ViewSchema) this is a
    /// summary of.
    pub name: ViewName,
    /// The result of [`ViewSchema::lazy()`](crate::schema::ViewSchema::lazy)
    /// for this view.
    pub lazy: bool,
    /// The result of [`ViewSchema::unique()`](crate::schema::ViewSchema::unique)
    /// for this view.
    pub unique: bool,
    /// The result of [`ViewSchema::version()`](crate::schema::ViewSchema::version)
    /// for this view.
    pub version: u64,
}
