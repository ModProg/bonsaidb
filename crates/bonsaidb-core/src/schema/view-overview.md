Views allow querying documents contained within [`Collection`s][Collection] in
an efficient manner.

The [`MapReduce::map()`] function is responsible for "mapping" data from the
stored document into the `Key` type. In the example below, the field `rank` is
being used as the View's [`Key`](View::Key) type, and the field `score` is being
used as the View's [`Value`](View::Value)

The [`MapReduce::reduce()`] function is responsible for "reducing" a list of
[`Value`s](View::Value) into a single value. The example below implements
`reduce()` to return an average value.

This enables [querying the view](crate::connection::View) in many powerful ways:

- [`db.view::<MyView>().query().await`](crate::connection::View::query): Return
   all entries in the view (without the associated documents).
- [`db.view::<MyView>().query_with_docs().await`](crate::connection::View::query_with_docs):
   Return all entries in the view, including the associated
   [`OwnedDocument`s](crate::document::OwnedDocument).
- [`db.view::<MyView>().query_with_collection_docs().await`](crate::connection::View::query_with_collection_docs):
   Return all entries in the view, including the associated
   [`CollectionDocument<T>`s](crate::document::CollectionDocument).
- [`db.view::<MyView>().reduce().await`](crate::connection::View::reduce):
   Returns the reduced value of the view query. For the example below, the
   result type of this call is `f32`.
- [`db.view::<MyView>().reduce_grouped().await`](crate::connection::View::reduce_grouped):
   Returns the reduced value of the view query, grouped by key. For the example
   below, the `value` returned will be an `f32` that is the average `score` of
   all documents with a matching `rank`.

All of the queries above can be filtered and customized by the methods available
on [`connection::View`](crate::connection::View).

For a more detailed walkthrough, see our [user guide's section on
Views](https://dev.bonsaidb.io/main/release/about/concepts/view.html).
