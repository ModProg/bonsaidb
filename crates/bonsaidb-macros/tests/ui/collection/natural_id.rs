use bonsaidb::core::schema::Collection;

#[derive(Collection)]
#[collection(name = "name", natural_id = Some(self.field))]
struct Test1 {
    #[natural_id]
    field: usize,
}

#[derive(Collection)]
#[collection(name = "name", natural_id = Some(self.field))]
struct Test2(#[natural_id] usize);

#[derive(Collection)]
#[collection(name = "name")]
struct Test3 {
    #[natural_id]
    field: usize,
    #[natural_id]
    field2: usize,
}

#[derive(Collection)]
#[collection(name = "name")]
struct Test4(#[natural_id] usize, #[natural_id] usize);

fn main() {}
