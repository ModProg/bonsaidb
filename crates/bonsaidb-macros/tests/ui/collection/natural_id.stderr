error: field marked as `natural_id` while `natural_id` expression is specified as well

         = help: remove `#[natural_id]` attribute on field
 --> tests/ui/collection/natural_id.rs:6:5
  |
6 |     #[natural_id]
  |     ^^^^^^^^^^^^^

error: `natural_id` expression is specified here
 --> tests/ui/collection/natural_id.rs:4:42
  |
4 | #[collection(name = "name", natural_id = Some(self.field))]
  |                                          ^^^^^^^^^^^^^^^^

error: field marked as `natural_id` while `natural_id` expression is specified as well

         = help: remove `#[natural_id]` attribute on field
  --> tests/ui/collection/natural_id.rs:12:14
   |
12 | struct Test2(#[natural_id] usize);
   |              ^^^^^^^^^^^^^

error: `natural_id` expression is specified here
  --> tests/ui/collection/natural_id.rs:11:42
   |
11 | #[collection(name = "name", natural_id = Some(self.field))]
   |                                          ^^^^^^^^^^^^^^^^

error: marked multiple fields as `natural_id`

         = note: currently only one field can be marked as `natural_id`
         = help: use `#[collection(natural_id=...)]` on the struct instead
  --> tests/ui/collection/natural_id.rs:19:5
   |
19 |     #[natural_id]
   |     ^^^^^^^^^^^^^

error: previous `natural_id`
  --> tests/ui/collection/natural_id.rs:17:5
   |
17 |     #[natural_id]
   |     ^^^^^^^^^^^^^

error: marked multiple fields as `natural_id`

         = note: currently only one field can be marked as `natural_id`
         = help: use `#[collection(natural_id=...)]` on the struct instead
  --> tests/ui/collection/natural_id.rs:25:35
   |
25 | struct Test4(#[natural_id] usize, #[natural_id] usize);
   |                                   ^^^^^^^^^^^^^

error: previous `natural_id`
  --> tests/ui/collection/natural_id.rs:25:14
   |
25 | struct Test4(#[natural_id] usize, #[natural_id] usize);
   |              ^^^^^^^^^^^^^
